src/service/price/PriceService.ts

```diff
@@ -1,7 +1,154 @@
 import { Service } from '@tsed/common';
+import { InternalServerError } from '@tsed/exceptions';
+import NodeCache from 'node-cache';
+import fetch from 'node-fetch';
+import { PriceData, TokenPrice } from '../../interface/TokenPrice';
+import { COINGECKO_URL, TOKENS } from '../../util/constants';
+import { getSushiswapPrice, getUniswapPrice } from '../../util/util';
+
+const priceCache = new NodeCache({ stdTTL: 300, checkperiod: 480 });
 
 /**
- * TODO: Implement pricing utility functions.
+ * API price oracle service. Uses CoinGecko as a source of truth for most
+ * tokens when possible, and TheGraph for AMM pairs when not available
+ * via CG. Prices are cached for 5 minutes at a time, but may live up to 8.
  */
 @Service()
-export class PriceService {}
+export class PriceService {
+	/**
+	 * Retrieve the USD price for a given token balance.
+	 * @param contract Token contract address.
+	 * @param balance Token balance to calculate price.
+	 */
+	async getUsdValue(contract: string, balance: number): Promise<number> {
+		const tokenPrice = await this.getTokenPriceData(contract);
+		return tokenPrice.usd * balance;
+	}
+
+	/**
+	 * Retrieve the ETH price for a given token balance.
+	 * @param contract Token contract address.
+	 * @param balance Token balance to calculate price.
+	 */
+	async getEthValue(contract: string, balance: number): Promise<number> {
+		const tokenPrice = await this.getTokenPriceData(contract);
+		return tokenPrice.eth * balance;
+	}
+
+	/**
+	 * Retrieve the price data for a given token in USD and ETH.
+	 * @param contract Token contract address.
+	 */
+	async getTokenPriceData(contract: string): Promise<TokenPrice> {
+		const cachedPrice: TokenPrice | undefined = priceCache.get(contract);
+		if (!cachedPrice) {
+			const priceData = await this.getPriceData();
+			return priceData[contract];
+		}
+		return cachedPrice;
+	}
+
+	/**
+	 * Retrieve all protocol token prices in both USD and ETH.
+	 */
+	async getPriceData(): Promise<PriceData> {
+		const [
+			badgerPrice,
+			diggPrice,
+			wbtcPrice,
+			uniBadgerWbtcPrice,
+			uniDiggWbtcPrice,
+			slpBadgerWbtcPrice,
+			slpDiggWbtcPrice,
+			slpEthWbtcPrice,
+			renBtcPrice,
+			sBtcPrice,
+			tBtcPrice,
+			wethPrice,
+			sushiPrice,
+		] = await Promise.all([
+			this.fetchContractPrice(TOKENS.BADGER),
+			this.fetchContractPrice(TOKENS.DIGG),
+			this.fetchContractPrice(TOKENS.WBTC),
+			getUniswapPrice(TOKENS.UNI_BADGER_WBTC),
+			getUniswapPrice(TOKENS.UNI_DIGG_WBTC),
+			getSushiswapPrice(TOKENS.SUSHI_BADGER_WBTC),
+			getSushiswapPrice(TOKENS.SUSHI_DIGG_WBTC),
+			getSushiswapPrice(TOKENS.SUSHI_ETH_WBTC),
+			this.fetchContractPrice(TOKENS.CRV_RENBTC),
+			this.fetchTokenPrice('sbtc'),
+			this.fetchTokenPrice('tbtc'),
+			this.fetchContractPrice(TOKENS.WETH),
+			this.fetchContractPrice(TOKENS.SUSHI),
+		]);
```

Comment:
It feels like there is a lot that could fail here. Maybe check the promise results for errors and handle it so that you're not just returning partial results?

Reply:
discussed offline